File: Data Structures & Algorithms/longest-palindromic-substring/submission-2.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def isPali(s):
            l, r = 0, len(s) - 1

            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True
        
        N = len(s)
        res = s[0]
        for l in range(N):
            for r in range(l+1, N):
                if isPali(s[l:r+1]):
                    res = s[l:r+1] if r - l + 1 > len(res) else res
        
        return res
        

File: Data Structures & Algorithms/anagram-groups/submission-2.java
class Solution {
    private boolean isAnagram(String s, String t) {
        int[] letters = new int[26];
        for(char c: s.toCharArray()){
            int index = (int)c - 'a';
            letters[index]++;
        }
        for(char c: t.toCharArray()){
            int index = (int)c - 'a';
            letters[index]--;
        }
        for(int i=0; i<26; i++) {
            if(letters[i]!= 0) return false;
        }
        return true;
    }
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> res = new ArrayList<>();
        for(int i = 0; i < strs.length; i++) {
            if(strs[i] != null) {
                List<String> curr = new ArrayList<>();
                curr.add(strs[i]);
                for(int j = i + 1 ; j < strs.length; j++) {
                    if(strs[j] != null && strs[j].length() == strs[i].length()) {
                        if(isAnagram(strs[i],strs[j])) {
                            curr.add(strs[j]);
                            strs[j] = null;
                        }
                    }
                }
                res.add(curr);
            }
        }
        return res;
    }
}
